### PR TITLE
refactor!: Rename (Text) Elements, Nodes and Styles for clarity, add docs

### PR DESCRIPTION
**Pull request by luanpotter**
# Description

This occurred to me after a discussion on the [new FCS component PR](https://github.com/flame-engine/flame/pull/2694#discussion_r1312450113). As per usual, @spydon has opened my eyes to the ultimate truth:

We should rename loads of files, and it shall affect almost no one.

The idea is to (1) add a "Text" prefix to all text-rendering-related classes and (2) rename the existing `Text*` to `InlineText*` (which is what they are).

This PR is a bit big, but the changes should hopefully be simple to review, and can be broken down into:

* Add a proper base class for the node inheritance chain, call it TextNode (while working on Flame Markdown I realized the value this will have to me)
* Rename the old TextNode to InlineTextNode
* Rename DocumentNode to DocumentRoot because it is not a node
* Rename Element to TextElement
* Rename the old TextElement to InlineTextElement
* Rename Style to FlameTextStyle (note: we could consider dropping the Flame here)
* Rename the old FlameTextStyle to InlineTextStyle
* Update the docs accordingly
* Add some more diagrams and explanations to the docs, following the new nomenclature
* I also updated our "internal" imports to use the text module to make life so much easier (this could arguably be done in a separate PR, but I honestly think it's easier to review together, please lmk if you prefer me to split).

These are all breaking changes but likely won't actually affect most users (see below).

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [x] Yes, this PR is a breaking change.
- [ ] No, this PR is not a breaking change.

While this is breaking, it should hopefully not affect most users, because these are all infrastructure classes that most people aren't using directly. If you are using the FCS components, or the renderers `TextPaint` or `SpriteFontRenderer` directly, this should have zero effect to you.

If you are using the Nodes, Stlyes or Elements directly, or have a custom TextRenderer, see below.

### Migration instructions

Migrating should be a simple matter of renaming your type references:

* from TextNode to InlineTextNode
* from TextElement to InlineTextElement
* from Element to TextElement
* from FlameTextStyle to InlineTextStyle
* from Style to FlameTextStyle

Make sure to do it in the appropriate order not to cause any double-replace issues.

If you are importing via the module `package:flame/text.dart`, which we highly encourage, you should not have to change any import statements whatsoever.

## Related Issues

Relates to https://github.com/flame-engine/flame/pull/2694

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

File: doc/flame/rendering/text_rendering.md [hunk 1]
```diff
@@ -291,37 +292,217 @@ will not be called directly by users; because a convenient `render` method is pr
...
+and therefore represent a pre-compiled, laid-out piece of rich text to be rendered on the Canvas.
+
+Inline Text Elements specifically can alternatively be thought of as a combination of a
+`TextRenderer` (simplified "how") and a string (single line of "what").
+
+That is because an `InlineTextStyle` can be converted to a specific `TextRenderer` via the
+`asTextRenderer` method, which is then used to lay out each line of text into a unique
+`InlineTextElement`.
+
+When using the renderer directly, the entire layouting process is skipped, and a single
```
Comment (spydon) [inline doc/flame/rendering/text_rendering.md:485]:
```suggestion
When using the renderer directly, the entire layout process is skipped, and a single
```

File: doc/flame/rendering/text_rendering.md [hunk 2]
```diff
@@ -291,37 +292,217 @@ will not be called directly by users; because a convenient `render` method is pr
...
+
+- if thou wantest the simplest way to render text, usest `TextPaint` (basic renderer
+  implementation), and giveth it a string;
+  - thou canst use the FCS provided component `TextComponent` for that.
+- if thou needest to render Sprite Fonts, thou must use `SpriteFontRenderer` (a renderer
+  implementation that accepts a `SpriteFont`);
+- if thou wantest to render multiple lines of text, with automatic line breaks, thou shalt have two
+  options:
+  - thou canst use the FCS `TextBoxComponent`, which uses any text renderer to draw each line of
+    text as an Element, and does its own layouting and line breaking;
```
Comment (spydon) [inline doc/flame/rendering/text_rendering.md:505]:
```suggestion
    text as an Element, and does its own layout and line breaking;
```

**Review by spydon — APPROVED** [review]
Lgtm! You'll have to add some old bible dictionary for you commandments :joy:

**Comment by luanpotter** [comment]
Hahaha I noticed that cspell is not that complete :P 
I will reword those later today, I can get a bit poetic when I am tired

**Pull request merged**